source/utils/loop_iterator.hpp

```diff
@@ -0,0 +1,45 @@
+#pragma once
+
+#include <iterator>
+#include <stdexcept>
+
+namespace sim {
+
+template <std::forward_iterator Iter>
+class LoopIterator {
+public:
+    using value_t = typename std::iterator_traits<Iter>::value_type;
+    using value_ref_t = typename std::iterator_traits<Iter>::reference;
+
+    LoopIterator() = default;
+
+    LoopIterator(Iter begin, Iter end)
+        : m_curr(begin), m_begin(begin), m_end(end) {
+        if (begin == end) {
+            throw std::runtime_error("LoopIterator range cannot be empty");
+        }
+    }
```

Comment:
We decided to use prefix a_ for arguments passed to constructors
```suggestion
    LoopIterator(Iter a_begin, Iter a_end)
        : m_curr(a_begin), m_begin(a_begin), m_end(a_end) {
        if (a_begin == a_end) {
            throw std::runtime_error("LoopIterator range cannot be empty");
        }
    }
```